### PR TITLE
Imironov/ref ngraph ctc gready decoder

### PR DESCRIPTION
**Pull request by iimironov**
Ticket 42645
Implement CTCGreedyDecoder reference

### CODE REVIEW

File: ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                                            const bool ctc_merge_repeated)
+            {
+                const auto batch_size = data_shape[0];
+                const auto class_count = data_shape[2];
+
+                CoordinateTransform out_transform = CoordinateTransform(out_shape);
+                CoordinateTransform data_transform = CoordinateTransform(data_shape);
+
+                std::vector<TF> tmp_out(shape_size(out_shape));
+                std::fill(tmp_out.begin(), tmp_out.end(), static_cast<TF>(-1.0));
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:46]: Why do you need this `fill` call?
I think if you need to create a vector with default value you can use this constructor:
```
explicit vector( size_type count,
                 const T& value = T(),
                 const Allocator& alloc = Allocator());
```

File: ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                        if (!(previous_class_index == max_class_ind && ctc_merge_repeated) &&
+                            max_class_ind < blank_index[0])
+                        {
+                            tmp_out[out_index++] = max_class_ind;
+                        }
+                        previous_class_index = max_class_ind;
+                    }
+                    out2[batch_ind] = seq_len;
+                }
+                std::copy(tmp_out.begin(), tmp_out.end(), out1);
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:70]: Why cannot we write the date immediately to `out1`?

File: ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                const auto class_count = data_shape[2];
+
+                CoordinateTransform out_transform = CoordinateTransform(out_shape);
+                CoordinateTransform data_transform = CoordinateTransform(data_shape);
+
+                std::vector<TCI> tmp_out(shape_size(out_shape), -1);
+
+                for (unsigned int batch_ind = 0; batch_ind < batch_size; batch_ind++)
+                {
+                    TI previous_class_index = static_cast<TI>(-1);
```
Comment (pelszkow) [inline ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:49]: `<limits>` are included so have you consider switch this to `std::numeric_limits<TI>::max()` instead `-1`.

File: ngraph/test/backend/ctc_greedy_decoder_seq_len.in.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    auto test_case = test::TestCase<TestEngine>(function);
+
+    test_case.add_input<float>({0.1f, 0.2f, 0.f, 0.4f, 0.3f, 0.f, 0.5f, 0.6f, 0.f});
+    test_case.add_input<int32_t>({2});
+    test_case.add_expected_output(Shape{N, T}, vector<int32_t>{1, 0, -1});
+    test_case.add_expected_output(Shape{N}, vector<int32_t>{2});
+
+    test_case.run();
+}
+/*
```
Comment (pelszkow) [inline ngraph/test/backend/ctc_greedy_decoder_seq_len.in.cpp:70]: Why this test are in comments?

File: ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                                            TCI* out1,
+                                            TSL* out2,
+                                            const Shape& data_shape,
+                                            const Shape& out_shape,
+                                            const bool ctc_merge_repeated)
+            {
+                const auto batch_size = data_shape[0];
+                const auto class_count = data_shape[2];
+
+                const CoordinateTransform out_transform{out_shape};
```
Comment (pelszkow) [inline ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:42]: @ilyachur , @vladislav-volkov, @iefode, @jane-intel, @jdanieck, @ggalieroc 
Is `CoordinateTransform` allowed now in reference implementations?

Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:42]: I propose to avoid it if it is possible.

Comment (iimironov) [reply ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:42]: Please give me examples or suggestion for replacing CoordinateTransform

Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/runtime/reference/ctc_greedy_decoder_seq_len.hpp:42]: A you can see in https://github.com/openvinotoolkit/openvino/pull/3633 we try to remove `CoordinateTransform` and calculate all in the fly. If you want to use something similar to `CooridianteTransform` look in to:
 https://github.com/openvinotoolkit/openvino/blob/bad4e97d9b9ea50146255acee9742572ff6baa47/ngraph/core/reference/src/runtime/reference/slice.cpp#L30

https://github.com/openvinotoolkit/openvino/blob/bad4e97d9b9ea50146255acee9742572ff6baa47/ngraph/core/reference/src/runtime/reference/reverse.cpp#L33

**Review by pelszkow — APPROVED** [review]

**Comment by ilyachur** [comment]
@iimironov Please resolve merge conflicts

**Review by ilyachur — APPROVED** [review]

**Pull request merged**